### PR TITLE
Jkmarx/data set owner transfer

### PR DESCRIPTION
**Pull request by jkmarx**
- Ref #2780 
- Ref #1947

### CODE REVIEW

File: refinery/core/test_views.py [hunk 1]
```diff
@@ -54,8 +54,8 @@ def setUp(self):
         )
 
         # Create Datasets
-        self.dataset = create_dataset_with_necessary_models(user=self.user)
-        self.dataset2 = create_dataset_with_necessary_models(user=self.user)
+        self.data_set = create_dataset_with_necessary_models(user=self.user)
```
Comment (jkmarx) [inline refinery/core/test_views.py]: Kept making typos so went ahead and updated the file to data_set vs dataset

File: refinery/core/test_views.py [hunk 2]
```diff
@@ -499,9 +554,162 @@ def test_is_filtered_data_set_returns_false_for_public_group(self):
         view_set = DataSetsViewSet()
         view_set.request = self.factory.get(self.url_root)
         view_set.request.user = SimpleLazyObject(lambda: self.user)
-        is_filtered = view_set.is_filtered_data_set(self.dataset, filter)
+        is_filtered = view_set.is_filtered_data_set(self.data_set, filter)
         self.assertFalse(is_filtered)
 
+    def test_send_tranfer_notification_email_corrent_users(self):
```
Comment (scottx611x) [inline refinery/core/test_views.py]: `tranfer` -> `transfer`
`corrent` -> `current`

File: refinery/core/test_views.py [hunk 3]
```diff
@@ -499,9 +554,162 @@ def test_is_filtered_data_set_returns_false_for_public_group(self):
...
+        self.assertIn(self.data_set.uuid, email.body)
+
+    def test_update_group_perms_remove_access(self):
+        new_owner_email = 'new_owner@fake.com'
+        new_owner = User.objects.create_user('NewOwner1', new_owner_email,
+                                             self.password)
+        self.data_set.share(ExtendedGroup.objects.public_group())
+        group_union = ExtendedGroup.objects.create(name="Group Union",
+                                                   is_public=True)
+        group_non_union = ExtendedGroup.objects.create(name="Group Non-Union",
```
Comment (scottx611x) [inline refinery/core/test_views.py]: I believe in Refinery there is only to be one public group. Maybe this isn't testing an exact scenario we'd encounter

File: refinery/core/test_views.py [hunk 4]
```diff
@@ -499,9 +554,162 @@ def test_is_filtered_data_set_returns_false_for_public_group(self):
...
+                                         group_non_union.extendedgroup.uuid)
+        )
+
+    def test_update_group_perms_retains_access(self):
+        new_owner_email = 'new_owner@fake.com'
+        new_owner = User.objects.create_user('NewOwner1', new_owner_email,
+                                             self.password)
+        self.data_set.share(ExtendedGroup.objects.public_group())
+        group_union = ExtendedGroup.objects.create(name="Group Union",
+                                                   is_public=True)
```
Comment (scottx611x) [inline refinery/core/test_views.py]: See above public group concern

File: refinery/core/test_views.py [hunk 5]
```diff
@@ -499,9 +554,162 @@ def test_is_filtered_data_set_returns_false_for_public_group(self):
...
+            groups.get('groups_with_access')[0].get('profile'),
+            'http://{}/groups/{}'.format(view_set.current_site,
+                                         group_union.extendedgroup.uuid)
+        )
+
+    def test_update_group_perms_retains_public(self):
+        new_owner_email = 'new_owner@fake.com'
+        new_owner = User.objects.create_user('NewOwner1', new_owner_email,
+                                             self.password)
+        group_union_public = ExtendedGroup.objects.public_group()
```
Comment (scottx611x) [inline refinery/core/test_views.py]: `group_union_public` -> `public_group` ?

File: refinery/core/test_views.py (hunk 3 above)
Comment (jkmarx) [reply refinery/core/test_views.py]: Unsure what you mean. The data set is shared with three groups (one being public) and the new owner only belongs to public and the union group. This test checks to see the data set is unshared with the non-union group.

File: refinery/core/views.py [hunk 6]
```diff
@@ -816,12 +817,33 @@ def delete(self, request, uuid):
...
+                new_owner_email = request.data.get('new_owner_email')
+                try:
+                    new_owner = User.objects.get(email=new_owner_email)
+                except Exception:
+                    return Response(uuid, status=status.HTTP_404_NOT_FOUND)
+
+                perm_groups = self.update_group_perms(new_owner)
+                self.send_tranfer_notification_email(current_owner,
+                                                     new_owner, perm_groups)
+                self.data_set.transfer_ownership(current_owner, new_owner)
```
[the remark this answers is not in the document]
Comment (jkmarx) [reply refinery/core/views.py]: Yes, the order is wrong. Must have jumbled it up when refactoring.

File: refinery/core/views.py [hunk 7]
```diff
@@ -833,9 +855,83 @@ def patch(self, request, uuid, format=None):
...
+            'groups_with_access': perm_groups.get('groups_with_access'),
+            'groups_without_access': perm_groups.get('groups_without_access')
+        }
+        email = EmailMessage(
+            subject,
+            temp_loader.render(context_dict),
+            to=[new_owner.email, old_owner.email]
+        )
+        email.send()
+        return email
```
[the remark this answers is not in the document]
Comment (jkmarx) [reply refinery/core/views.py]: It's useful for testing.

File: refinery/core/test_views.py [hunk 8]
```diff
@@ -197,15 +198,108 @@ def test_dataset_delete_not_found(self):
...
+        )
+        patch_request.user = self.user
+        force_authenticate(patch_request, user=self.user)
+        self.view(patch_request, self.data_set.uuid)
+        self.assertTrue(mock_update.called)
+        self.assertTrue(mock_email.called)
+
+    @mock.patch('core.views.DataSetsViewSet.update_group_perms',
+                side_effect=RuntimeError)
+    def test_dataset_patch_fails_and_rollback_owner(self, mock_update):
```
Comment (jkmarx) [inline refinery/core/test_views.py]: @scottx611x These are the additional tests

File: refinery/core/views.py [hunk 9]
```diff
@@ -816,12 +818,41 @@ def delete(self, request, uuid):
...
+            current_owner = self.data_set.get_owner()
+            if request.data.get('transfer_data_set') and current_owner == \
+                    request.user:
+                new_owner_email = request.data.get('new_owner_email')
+                try:
+                    new_owner = User.objects.get(email=new_owner_email)
+                except Exception:
+                    return Response(uuid, status=status.HTTP_404_NOT_FOUND)
+
+                try:
```
Comment (jkmarx) [inline refinery/core/views.py]: @scottx611x Error handling and rollback

File: refinery/core/test_views.py (hunk 8 above)
Comment (jkmarx) [reply refinery/core/test_views.py]: I did confirm the unsharing

File: refinery/core/views.py [hunk 10]
```diff
@@ -832,8 +833,16 @@ def patch(self, request, uuid, format=None):
...
-                self.data_set.transfer_ownership(current_owner, new_owner)
-                perm_groups = self.update_group_perms(new_owner)
+                try:
+                    with transaction.atomic():
+                        self.data_set.transfer_ownership(current_owner,
+                                                         new_owner)
+                        perm_groups = self.update_group_perms(new_owner)
+                except Exception as e:
+                    return Response(
+                        e, status=status.HTTP_500_INTERNAL_SERVER_ERROR
```
Comment (scottx611x) [inline refinery/core/views.py]: Thinking again, this would be more appropriate as an error in the 400's. Its weird to explicitly send back a 500

Comment (jkmarx) [reply refinery/core/views.py]: @scottx611x Suggestions?
http://www.django-rest-framework.org/api-guide/status-codes/#client-error-4xx

Comment (jkmarx) [reply refinery/core/views.py]: It's not really a bad request is it?

Comment (scottx611x) [reply refinery/core/views.py]: Maybe `HTTP_412_PRECONDITION_FAILED` ?

Comment (hackdna) [reply refinery/core/views.py]: Good reference for HTTP response status codes: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

**Comment by jkmarx** [comment]
@scottx611x After reading @hackdna posts, I think the 412 is appropriate. Were there any other changes?

**Comment by scottx611x** [comment]
@jkmarx Its been a while since I looked at the whole PR. I'll take a peek and let you know

**Pull request merged**